### PR TITLE
Expanded Sinon's createStubInstance to allow a TType generic

### LINKED ISSUE #16728: Sinon's createStubInstance should return a mapped type
Instead of `any`, there should be more expressive typings explaining that it returns a mapped type of the constructor or object it's passed.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Fixes #16728.

### CODE REVIEW

**Comment by nicojs** [comment]
This looks awesome! Does it work without the type argument as well? :+1:

**Comment by JoshuaKGoldberg** [comment]
Yup! As per the tests:

```typescript
class TestCreateStubInstance {
    someTestMethod(testArg: string) { }
}

sinon.createStubInstance(TestCreateStubInstance).someTestMethod('some argument');
sinon.createStubInstance<TestCreateStubInstance>(TestCreateStubInstance).someTestMethod('some argument');
```

**Comment by nicojs** [comment]
Ah! Nice. Didn't see that one. I'm extremely excited about this pr, as i've been creating this function and mapped type in code often times myself. Didn't even know that `createStubInstance` exists :)

**Comment by JoshuaKGoldberg** [comment]
Oh geez, forgot to bump for 2.2... will do.

**Comment by mhegazy** [comment]
@JoshuaKGoldberg can you please update the PR.

File: types/sinon/index.d.ts [hunk 1]
```diff
@@ -495,10 +495,47 @@ declare namespace Sinon {
...
     }
+
+    /**
+     * Stubbed type of an object with members replaced by stubs.
+     * 
+     * @type TType   Type being stubbed.
+     */
+    type StubbableType<TType> = {
+        new?(): TType;
+        new?(...args: any[]): TType;
```
Comment (andy-ms) [inline types/sinon/index.d.ts]: Don't think both overloads are needed here.

**Comment by JoshuaKGoldberg** [comment]
I don't understand why `sinon-test/sinon-test-tests.ts` is failing with `Package compiles in TypeScript 2.3 but not in 2.0.`. Both it and `sinon` have `// TypeScript Version: 2.3` in their headers. Am I missing a step somewhere?

File: types/sinon-test/index.d.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@
 // Project: https://github.com/sinonjs/sinon-test
 // Definitions by: Francis Saul <https://github.com/mummybot>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript: 2.3
```
Comment (andy-ms) [inline types/sinon-test/index.d.ts]: `// TypeScript Version: 2.3`. (Another case of Microsoft/dtslint#43.)

**Comment by JoshuaKGoldberg** [comment]
Oh.

**Comment by andy-ms** [comment]
Thanks!

**Pull request merged**